File: python/evm_state/host_recovery.py
"""Explicit same-machine recovery for hostname-bound prototype runs.

This cannot establish where an old hostname ran. The operator must already know
this is the original machine. It preserves every original identity/cursor/prefix
and binds the local recovery records to the persistent OS identity.
"""
from contextlib import ExitStack
import hashlib
import json
import os
from pathlib import Path

from .bootstrap import load_prefix
from .checkpoint import connect_like
from .cursor import load_progress, validate
from .files import atomic_json, file_lock
from .host import recovery_record
from .proof import VerificationError


def rebind(client, directory, previous_host):
    directory = Path(directory).resolve()
    if not previous_host or previous_host.startswith("machine-sha256:"):
        raise VerificationError("recovery requires the exact legacy hostname")
    run = json.loads((directory / "run.json").read_text())
    identity = run["identity"]
    checkpoints = connect_like(client, identity["checkpoint_database"])
    checkpoint_uuid = checkpoints.one("SELECT toString(uuid) AS id FROM system.databases WHERE name={db:String}",
                                      {"db": checkpoints.database})["id"]
    control_dir = Path(os.environ.get("EVM_STATE_HOME", "localdata/control")).resolve() / checkpoint_uuid
    with ExitStack() as locks:
        for path in [directory / "bootstrap_replay.lock", directory / "run.lock", directory / "source_readers.lock",
                     control_dir / "initialize.lock", control_dir / "writer.lock", control_dir / "readers.lock"]:
            locks.enter_context(file_lock(path, exclusive=True, blocking=False))
        # Compare both sides before creating either sidecar. Recovery is
        # repeatable if interrupted between the two atomic file writes.
        if json.loads((directory / "run.json").read_text()) != run:
            raise VerificationError("run changed during host recovery")
        owner = client.one("SELECT run_id,identity FROM _evm_state_run")
        db_uuid = client.one("SELECT toString(uuid) AS id FROM system.databases WHERE name={db:String}",
                             {"db": client.database})["id"]
        if (run.get("phase") != "prepared" or identity.get("format_version") != 3
                or identity.get("host") != previous_host or identity.get("state_directory") != str(directory)
                or identity.get("database") != client.database or identity.get("http_url") != client.url
                or run.get("database_uuid") != db_uuid or owner["run_id"] != run["run_id"]
                or json.loads(owner["identity"]) != identity):
            raise VerificationError("legacy native ownership does not match")
        if (hashlib.sha256((directory / "package.spkg").read_bytes()).hexdigest() != identity["package_sha256"]
                or (directory / "meta" / f"{client.database}_schema_hash.txt").read_text().strip() != run["schema_hash"]):
            raise VerificationError("frozen package or schema differs")
        binding = json.loads((control_dir / "binding.json").read_text())
        control_owner = checkpoints.one("SELECT control_id,binding FROM _evm_checkpoint_control")
        if (binding.get("host") != previous_host or binding.get("database_uuid") != checkpoint_uuid
                or binding.get("directory") != str(control_dir)
                or control_owner["control_id"] != binding["control_id"]
                or json.loads(control_owner["binding"]) != binding
                or (control_dir / "initialized").read_text() != binding["control_id"]):
            raise VerificationError("legacy checkpoint ownership does not match")
        progress = load_progress(client, run, directory)
        cursor = validate(client, run, (directory / "cursor.txt").read_text().strip())
        if cursor["block"]["number"] < progress["position"]["block"]["number"]:
            raise VerificationError("native cursor regressed behind durable progress")
        prefix = load_prefix(client, directory, run)
        for record, path in [(identity, directory), (binding, control_dir)]:
            target = path / "host-rebinding.json"
            expected = recovery_record(record, path)
            if target.exists() and json.loads(target.read_text()) != expected:
                raise VerificationError("existing host recovery belongs to different state or machine")
        for record, path in [(identity, directory), (binding, control_dir)]:
            target = path / "host-rebinding.json"
            if not target.exists():
                atomic_json(target, recovery_record(record, path))
        return {"run_id": run["run_id"], "rebound": True, "cursor_block": cursor["block"],
                "prefix_block": prefix["header"] if prefix else None,
                "qualification": "Operator-confirmed original machine; original identities and state preserved."}

File: crates/evm-state/src/host_recovery.rs
//! Explicit same-machine recovery for legacy hostname-bound source/controller pairs.
use crate::{
    bootstrap,
    ch::{params, uint, ClickHouse},
    cursor,
    files::{atomic_json, file_lock, resolve},
    host,
    proof::string,
    trie_qualification::file_hash,
};
use anyhow::{ensure, Context, Result};
use serde_json::{json, Value};
use std::{fs, path::Path};

/// The operator must attest that this is the original machine. A legacy
/// hostname cannot establish that fact cryptographically. Original state and
/// identities stay byte-for-byte intact; only checked recovery sidecars are added.
pub fn rebind(client: &ClickHouse, directory: &Path, previous_host: &str) -> Result<Value> {
    ensure!(
        !previous_host.is_empty() && !previous_host.starts_with("machine-sha256:"),
        "recovery requires the exact legacy hostname"
    );
    let directory = resolve(directory)?;
    let run: Value = serde_json::from_slice(&fs::read(directory.join("run.json"))?)?;
    let identity = &run["identity"];
    let checkpoints = client.with_database(string(identity, "checkpoint_database")?)?;
    let checkpoint_uuid = checkpoints.one(
        "SELECT toString(uuid) AS id FROM system.databases WHERE name={db:String}",
        &params(json!({"db":checkpoints.database}))?,
    )?["id"]
        .clone();
    let id = uuid::Uuid::parse_str(
        checkpoint_uuid
            .as_str()
            .context("invalid checkpoint database UUID")?,
    )?;
    ensure!(
        !id.is_nil() && json!(id.to_string()) == checkpoint_uuid,
        "checkpoint coordination requires an Atomic ClickHouse database"
    );
    let control_dir = resolve(&client.control_home)?.join(id.to_string());
    let mut locks = Vec::new();
    for path in [
        directory.join("bootstrap_replay.lock"),
        directory.join("run.lock"),
        directory.join("source_readers.lock"),
        control_dir.join("initialize.lock"),
        control_dir.join("writer.lock"),
        control_dir.join("readers.lock"),
    ] {
        locks.push(file_lock(&path, true, false)?);
    }
    ensure!(
        serde_json::from_slice::<Value>(&fs::read(directory.join("run.json"))?)? == run,
        "run changed during host recovery"
    );
    let owner = client.one(
        "SELECT run_id,identity FROM _evm_state_run",
        &Default::default(),
    )?;
    let database = client.one(
        "SELECT toString(uuid) AS id FROM system.databases WHERE name={db:String}",
        &params(json!({"db":client.database}))?,
    )?;
    ensure!(
        run["phase"] == "prepared"
            && identity["format_version"] == 3
            && identity["host"] == previous_host
            && identity["state_directory"]
                == directory.to_str().context("invalid state directory")?
            && identity["database"] == client.database
            && identity["http_url"] == client.url
            && run["database_uuid"] == database["id"]
            && owner["run_id"] == run["run_id"]
            && serde_json::from_str::<Value>(string(&owner, "identity")?)? == *identity,
        "legacy native ownership does not match"
    );
    ensure!(
        file_hash(&directory.join("package.spkg"))? == string(identity, "package_sha256")?
            && fs::read_to_string(
                directory
                    .join("meta")
                    .join(format!("{}_schema_hash.txt", client.database))
            )?
            .trim()
                == string(&run, "schema_hash")?,
        "frozen package or schema differs"
    );
    let binding: Value = serde_json::from_slice(&fs::read(control_dir.join("binding.json"))?)?;
    let control_owner = checkpoints.one(
        "SELECT control_id,binding FROM _evm_checkpoint_control",
        &Default::default(),
    )?;
    ensure!(
        binding["host"] == previous_host
            && binding["database_uuid"] == checkpoint_uuid
            && binding["directory"]
                == control_dir
                    .to_str()
                    .context("invalid controller directory")?
            && control_owner["control_id"] == binding["control_id"]
            && serde_json::from_str::<Value>(string(&control_owner, "binding")?)? == binding
            && fs::read_to_string(control_dir.join("initialized"))?
                == string(&binding, "control_id")?,
        "legacy checkpoint ownership does not match"
    );
    let progress = cursor::load_progress(client, &run, &directory)?;
    let native = cursor::validate(
        client,
        &run,
        fs::read_to_string(directory.join("cursor.txt"))?.trim(),
    )?;
    ensure!(
        native.block.number >= uint(&progress["position"]["block"]["number"])?,
        "native cursor regressed behind durable progress"
    );
    let prefix = bootstrap::load_prefix(client, &directory, Some(&run))?;
    let machine = host::machine_id()?;
    let expected = [
        (
            directory.join("host-rebinding.json"),
            host::recovery_record_for(identity, &directory, &machine)?,
        ),
        (
            control_dir.join("host-rebinding.json"),
            host::recovery_record_for(&binding, &control_dir, &machine)?,
        ),
    ];
    // Validate both sides before writing either. A single matching sidecar from
    // an interrupted attempt can safely resume the second atomic write.
    for (path, record) in &expected {
        ensure!(
            !path.try_exists()? || serde_json::from_slice::<Value>(&fs::read(path)?)? == *record,
            "existing host recovery belongs to different state or machine"
        );
    }
    for (path, record) in &expected {
        if !path.try_exists()? {
            atomic_json(path, record, false)?;
        }
    }
    Ok(
        json!({"run_id":run["run_id"],"rebound":true,"cursor_block":native.block,"prefix_block":prefix.map(|p|p["header"].clone()),"qualification":"Operator-confirmed original machine; original identities and state preserved."}),
    )
}
